packages/rest/src/session/ConnectionPropsForSessCfg.ts

```diff
@@ -0,0 +1,285 @@
+     *          basePath: commandParameters.arguments.basePath
+     *      };
+     *      const connectableSessCfg = await ConnectionPropsForSessCfg.addPropsOrPrompt<ISession>(
+     *           sessCfg, commandParameters.arguments
+     *      );
+     *      mySession = new Session(connectableSessCfg);
+     *
+     * @returns A session configuration object with connection information
+     *          added to the initialSessCfg. Its intended use is for our
+     *          caller to create a session for a REST Client.
+     */
+    public static async addPropsOrPrompt<T>(
+        initialSessCfg: T,
+        cmdArgs: ICommandArguments,
+        options: IOptionsForAddConnProps = {}
+    ): Promise<T> {
+        const impLogger = Logger.getImperativeLogger();
+
+        const optionDefaults: IOptionsForAddConnProps = {
+            requestToken: false,
+            doPrompting: true,
+            defaultTokenType: SessConstants.TOKEN_TYPE_JWT
+        };
+
+        // override our defaults with what our caller wants.
+        const optsToUse = {...optionDefaults, ...options};
+
+        // initialize session config object with what our caller supplied
+        const finalSessCfg: any = initialSessCfg;
+
+        /* Override properties from our caller's initialSessCfg
+         * with any values from the command line.
+         */
+        if (ConnectionPropsForSessCfg.propHasValue(cmdArgs.host)) {
+            finalSessCfg.hostname = cmdArgs.host;
+        }
+        if (ConnectionPropsForSessCfg.propHasValue(cmdArgs.port)) {
+            finalSessCfg.port = cmdArgs.port;
+        }
+        if (ConnectionPropsForSessCfg.propHasValue(cmdArgs.user)) {
+            finalSessCfg.user = cmdArgs.user;
+        }
+        if (ConnectionPropsForSessCfg.propHasValue(cmdArgs.password)) {
+            finalSessCfg.password = cmdArgs.password;
+        }
+        if (optsToUse.requestToken) {
+            // deleting any tokenValue, ensures that basic creds are used to authenticate and get token
+            delete finalSessCfg.tokenValue;
+        } else {
+            if (ConnectionPropsForSessCfg.propHasValue(cmdArgs.tokenValue)) {
+                finalSessCfg.tokenValue = cmdArgs.tokenValue;
+            }
+        }
+
+        // if our caller permits, prompt for host and port as needed
+        if (optsToUse.doPrompting) {
+            if (ConnectionPropsForSessCfg.propHasValue(finalSessCfg.hostname) === false) {
+                let answer = "";
+                while (answer === "") {
+                    answer = await CliUtils.promptWithTimeout(
+                        "Enter the host name of your service: "
+                    );
+                    if (answer === null) {
+                        throw new ImperativeError({msg: "We timed-out waiting for host name."});
+                    }
+                }
+                finalSessCfg.hostname = answer;
+            }
```

Comment:
It'd be nice if we could reuse this entire block of code since it's kind of the same for every option, with the exception of string messages (which should ideally be extracted somewhere for localization purposes, but I'm having a hard time believing we will ever get to localizing the framework)

No changes requested, just whining! 😋

Reply:
If we do a bunch more prompting in a future effort, maybe we should replace (or wrapper) CliUtils.promptWithTimeout with a function that takes a name for the data value, forms the prompt string, does the while loop, and returns the answer or issues the timeout error.